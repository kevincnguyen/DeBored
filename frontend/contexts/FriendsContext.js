import React, { createContext, useContext, useState, useMemo } from "react";

const FriendsContext = createContext();

/*
 * Provider that manages the friends state
 */
export const FriendsProvider = ({ children }) => {
  const [friends, setFriends] = useState([]);

  // Function that updates the friends information
  const updateFriends = (newFriends) => {
    setFriends(newFriends);
  };

  // Memoize the context value to prevent unnecessary renders
  const contextValue = useMemo(
    () => ({
      friends,
      updateFriends,
    }),
    [friends]
  );

  return (
    <FriendsContext.Provider value={contextValue}>
      {children}
    </FriendsContext.Provider>
  );
};

// Hook to easily use the friends context
export const useFriends = () => {
  const context = useContext(FriendsContext);

  if (!context) {
    throw new Error("useFriends must be used within a FriendsContext");
  }

  return context;
};
