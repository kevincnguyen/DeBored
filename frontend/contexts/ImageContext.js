import React, { createContext, useContext, useState, useMemo } from "react";

const ImageContext = createContext();

/*
 * Provider that manages the image state
 */
export const ImageProvider = ({ children }) => {
  const [imageKey, setImageKey] = useState(0);

  // Function that updates the image information
  const updateImageKey = () => {
    setImageKey((prev) => prev + 1);
  };

  // Memoize the context value to prevent unnecessary renders
  const contextValue = useMemo(
    () => ({
      imageKey,
      updateImageKey,
    }),
    [imageKey]
  );

  return (
    <ImageContext.Provider value={contextValue}>
      {children}
    </ImageContext.Provider>
  );
};

// Hook to easily use the image context
export const useImage = () => {
  const context = useContext(ImageContext);

  if (!context) {
    throw new Error("useImage must be used within a ImageProvider");
  }

  return context;
};
