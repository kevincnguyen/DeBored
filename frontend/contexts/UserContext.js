import React, { createContext, useContext, useState, useMemo } from "react";

const UserContext = createContext();

/*
 * Provider that manages the user state
 */
export const UserProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  // Function that updates the user information
  const updateUser = (newUser) => {
    setUser(newUser);
  };

  // Memoize the context value to prevent unnecessary renders
  const contextValue = useMemo(
    () => ({
      user,
      updateUser,
    }),
    [user]
  );

  return (
    <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>
  );
};

// Hook to easily use the user context
export const useUser = () => {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error("useUser must be used within a UserProvider");
  }

  return context;
};
