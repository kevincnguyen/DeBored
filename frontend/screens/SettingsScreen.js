import React, { useState } from "react";
import {
  View,
  Text,
  ScrollView,
  Image,
  StyleSheet,
  Keyboard,
  Alert,
  ActivityIndicator,
  KeyboardAvoidingView,
} from "react-native";
import { Button, TextInput } from "react-native-paper";
import * as ImagePicker from "expo-image-picker";
import * as FileSystem from "expo-file-system";

import { useUser } from "../contexts/UserContext";
import { useImage } from "../contexts/ImageContext";
import DismissKeyboard from "../components/Utils/DismissKeyboard";

/*
 * The profile settings screen that allows users to update information.
 */
const SettingsScreen = ({ handleImageRender }) => {
  const { user, updateUser } = useUser();
  const { updateImageKey } = useImage();
  const [loading, setLoading] = useState(false);
  const [image, setImage] = useState(null);
  const [name, setName] = useState(user.name);
  const [bio, setBio] = useState(user.bio);
  const [email, setEmail] = useState(user.email);
  const [password, setPassword] = useState(user.password);
  const [confirmPassword, setConfirmPassword] = useState("");
  const [hidePassword, setHidePassword] = useState(true);
  const [hideConfirmPassword, setHideConfirmPassword] = useState(true);
  const [phone, setPhone] = useState(user.phone);
  const [instagram, setInstagram] = useState(user.instagram);
  const [twitter, setTwitter] = useState(user.twitter);
  const [facebook, setFacebook] = useState(user.facebook);

  const handleChangePicture = async () => {
    let result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [4, 3],
      quality: 0.7,
      height: 800,
      width: 800,
    });
    if (!result.canceled) {
      setImage(result.assets[0].uri);
    } else {
      setImage(null);
    }
  };

  const handleSubmit = async () => {
    Keyboard.dismiss();
    // check if all required fields are filled
    if (name === "" || password === "") {
      Alert.alert(
        "Error",
        "Please fill out the name and/or password fields",
        [
          {
            text: "OK",
            onPress: () => console.log("OK pressed"),
          },
        ],
        { cancelable: false }
      );
      return;
    }
    // check if passwords match
    if (confirmPassword !== "" && password !== confirmPassword) {
      Alert.alert(
        "Error",
        "Passwords do not match",
        [
          {
            text: "OK",
            onPress: () => console.log("OK pressed"),
          },
        ],
        { cancelable: false }
      );
      return;
    }
    // Convert to Base64 if defined
    let imageBase64;
    if (image) {
      imageBase64 = await FileSystem.readAsStringAsync(image, {
        encoding: "base64",
      });
    }

    // Update S3 and MongoDB
    setLoading(true);
    fetch(
      "https://5cfqsu7tzg.execute-api.us-west-2.amazonaws.com/default/EditInfo",
      {
        method: "POST",
        body: JSON.stringify({
          id: user._id,
          name: name,
          bio: bio,
          password: password,
          image: imageBase64,
          instagram: instagram,
          twitter: twitter,
          facebook: facebook,
          phone: phone,
        }),
      }
    )
      .then((response) => {
        if (response.status === 200) {
          return response.json();
        } else {
          console.log(user);
          throw new Error("Invalid profile update");
        }
      })
      .then((response) => {
        console.log(`response: ${JSON.stringify(response)}`);
        updateUser({
          ...user,
          name: name,
          bio: bio,
          password: password,
          instagram: instagram,
          twitter: twitter,
          facebook: facebook,
          phone: phone,
        });
        updateImageKey();
        setImage(null);
        setLoading(false);
      })
      .catch((err) => {
        setLoading(false);
        console.log(err);
        Alert.alert(
          "Error",
          "Invalid profile update, please try again.",
          [
            {
              text: "OK",
              onPress: () => console.log("OK pressed"),
            },
          ],
          { cancelable: false }
        );
      });
  };

  return (
    <KeyboardAvoidingView
      style={styles.keyboardAvoid}
      behavior="padding"
      enabled
      keyboardVerticalOffset={100}
    >
      <DismissKeyboard>
        <View style={styles.container}>
          <ScrollView
            contentContainerStyle={styles.scroll}
            showsVerticalScrollIndicator={false}
          >
            <Image
              source={{ uri: image || user.profilePicURL }}
              style={styles.picture}
            />
            <Button
              mode="contained-tonal"
              title="Submit"
              onPress={handleChangePicture}
            >
              Edit picture
            </Button>
            <View style={styles.inputs}>
              <Text style={styles.header}>Profile</Text>
              <TextInput
                mode="outlined"
                label="Name"
                value={name}
                onChangeText={(name) => setName(name)}
              />
              <TextInput
                mode="outlined"
                label="Bio"
                multiline={true}
                numberOfLines={4}
                value={bio}
                onChangeText={(bio) => setBio(bio)}
              />
              <TextInput
                mode="outlined"
                label="Email"
                value={email}
                onChangeText={(email) => setEmail(email)}
                disabled={true}
              />
              <TextInput
                mode="outlined"
                label="Password"
                value={password}
                onChangeText={(password) => setPassword(password)}
                secureTextEntry={hidePassword}
                right={
                  <TextInput.Icon
                    icon={hidePassword ? "eye-off" : "eye"}
                    onPress={() => {
                      Keyboard.dismiss();
                      setHidePassword(!hidePassword);
                    }}
                    color="black"
                  />
                }
              />
              <TextInput
                mode="outlined"
                label="Confirm Password"
                value={confirmPassword}
                onChangeText={(confirmPassword) =>
                  setConfirmPassword(confirmPassword)
                }
                secureTextEntry={hideConfirmPassword}
                right={
                  <TextInput.Icon
                    icon={hideConfirmPassword ? "eye-off" : "eye"}
                    onPress={() => {
                      Keyboard.dismiss();
                      setHideConfirmPassword(!hideConfirmPassword);
                    }}
                    color="black"
                  />
                }
              />
            </View>
            <View style={styles.inputs}>
              <Text style={styles.header}>Contact Information</Text>
              <TextInput
                mode="outlined"
                label="Phone Number"
                value={phone}
                onChangeText={(phone) => setPhone(phone)}
              />
              <TextInput
                mode="outlined"
                label="Instagram Username"
                value={instagram}
                onChangeText={(instagram) => setInstagram(instagram)}
              />
              <TextInput
                mode="outlined"
                label="Twitter Username"
                value={twitter}
                onChangeText={(twitter) => setTwitter(twitter)}
              />
              <TextInput
                mode="outlined"
                label="Facebook Link"
                value={facebook}
                onChangeText={(facebook) => setFacebook(facebook)}
              />
              {loading ? (
                <ActivityIndicator size="large" />
              ) : (
                <Button
                  mode="contained"
                  title="Submit"
                  onPress={handleSubmit}
                  style={styles.button}
                >
                  Save Changes
                </Button>
              )}
            </View>
          </ScrollView>
        </View>
      </DismissKeyboard>
    </KeyboardAvoidingView>
  );
};

const styles = StyleSheet.create({
  keyboardAvoid: {
    flex: 1,
    flexDirection: "column",
    justifyContent: "center",
  },
  container: {
    flex: 1,
    paddingHorizontal: 10,
    backgroundColor: "#FFFFFF",
  },
  scroll: {
    flexGrow: 1,
    alignItems: "center",
  },
  inputs: {
    width: "95%",
    flexDirection: "column",
    gap: 10,
    marginBottom: 30,
  },
  picture: {
    height: 120,
    width: 120,
    borderRadius: 60,
    marginBottom: 10,
    marginTop: 20,
  },
  header: {
    fontWeight: "bold",
    marginLeft: 3,
    marginTop: 15,
    color: "#7845AC",
  },
  button: {
    marginTop: 15,
  },
});

export default SettingsScreen;
