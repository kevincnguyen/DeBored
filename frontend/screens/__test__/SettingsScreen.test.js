import React from "react";
import renderer from "react-test-renderer";
import SettingsScreen from "../SettingsScreen";
import { useUser } from "../../contexts/UserContext";

// Mock the user context
jest.mock("../../contexts/UserContext", () => ({
  useUser: jest.fn(),
}));

// Mock the react-native-paper components
jest.mock("react-native-paper", () => ({
  Button: "Button",
  TextInput: "TextInput",
}));

test("renders SettingsScreen component correctly", () => {
  useUser.mockImplementation(() => ({
    user: { name: "", bio: "", email: "" },
  }));
  const tree = renderer.create(<SettingsScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("contains all input fields with correct labels", () => {
  const component = renderer.create(<SettingsScreen />);
  const textInputs = component.root.findAllByType("TextInput");
  expect(textInputs.length).toBe(5); // 5 inputs: name, bio, email, password, confirm password
});

test("contains buttons for changing picture and saving changes", () => {
  const component = renderer.create(<SettingsScreen />);
  const buttons = component.root.findAllByType("Button");
  expect(buttons.length).toBe(2); // 2 buttons: edit picture and save changes
});
