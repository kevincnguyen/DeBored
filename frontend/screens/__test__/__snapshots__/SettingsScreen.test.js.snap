// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SettingsScreen component correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "padding": 10,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "alignItems": "center",
        "flexGrow": 1,
      }
    }
  >
    <View>
      <Image
        source={1}
        style={
          {
            "borderRadius": 60,
            "height": 120,
            "marginBottom": 10,
            "width": 120,
          }
        }
      />
      <Button
        mode="contained-tonal"
        onPress={[Function]}
        title="Submit"
      >
        Edit picture
      </Button>
      <View
        style={
          {
            "flexDirection": "column",
            "gap": 10,
            "marginVertical": 20,
            "width": "100%",
          }
        }
      >
        <TextInput
          label="Name"
          mode="outlined"
          onChangeText={[Function]}
          value=""
        />
        <TextInput
          label="Bio"
          mode="outlined"
          multiline={true}
          numberOfLines={4}
          onChangeText={[Function]}
          value=""
        />
        <TextInput
          label="Email"
          mode="outlined"
          onChangeText={[Function]}
          value=""
        />
        <TextInput
          label="Password"
          mode="outlined"
          onChangeText={[Function]}
          value=""
        />
        <TextInput
          label="Confirm Password"
          mode="outlined"
          onChangeText={[Function]}
          value=""
        />
        <Button
          mode="contained"
          onPress={[Function]}
          title="Submit"
        >
          Save Changes
        </Button>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
